Use variable length primitives in raw serial form of cache key/value objects

AFaust opened the pull request:
### CHECKLIST

We will not consider a PR until the following items are checked off--thank you!

- [x] There aren't existing pull requests attempting to address the issue mentioned here
- [x] Submission developed in a feature branch--not master

### CONVINCING DESCRIPTION

This PR adds support for variable length integer use in the raw serial form of cache key / value objects. Variable length integers are implemented by writing values in incremental byte fragments with the top bit of each fragment except for the last indicating whether another fragment follows. This is used for database IDs, String lengths, content file sizes and regular node properties (within the supported value ranges). Additionally, this PR also includes the reclassification of the efficiency tests as "expensive tests" to focus the automated / regular builds on the correctness aspect.

### RELATED INFORMATION

Fixes #36
Fixes #42

# BE WARNED

- If any of the checklist items are missed, incomplete or invalid we will not accept the PR
- If you are not responsive to feedback on your PR we will not accept the PR
- If we do not accept your PR we will provide feedback on the PR indicating why
- After some time, we will close PRs that have not been accepted

When we decline or close your PR. You are encouraged to address the concerns outlined in the
PR and then re-submit it. We want contributions. We also need for them to be of high quality
and high value and to not take undue resources away from the features and enhancements the
core team is working on. Thanks for understanding!

AFaust committed:
Variable length primitives (#36)

AFaust committed:
Rename feature, document and minor key optimisations (#36)

AFaust committed:
Classify efficiency tests as expensive (#42)

AFaust commented:
Updated memory and throughput benchmark is pending and will be done when all binary serialiser enhancement issues have been processed. I will based the PRs of the various features off of each other, meaning they will be cascading PRs that should be processed / merged in sequence. For the time being, this PR should be left unmerged until the final issue / the benchmark has been done.

AFaust committed:
Refactor to reduce CPU overhead (#36)

- minor reduction in memory savings as longs are now always written with
  short fragments (only affects the very smallest values / those with odd
  number of bytes before)

andreaskring commented on repository/src/main/java/org/aldica/repo/ignite/binary/NodePropertiesBinarySerializer.java:
```diff
@@ -315,13 +313,31 @@ else if (value instanceof Boolean)
         {
-            rawWriter.writeByte(TYPE_INTEGER);
-            rawWriter.writeInt((Integer) value);
+            final int val = ((Integer) value).intValue();
```
These code blocks (lines 316 - 326, 330 - 340 and 359 - 369) seems to be almost identical. Maybe the logic in these blocks could be refactored into a generic method (using appropriate generic types and arguments)?

AFaust replied:
330-340 and 359-369 could potentially be extracted into a common method, but 316-326 differs with regards to value/parameter types, so would not be applicable, unless we are fine with suffering the additional cost of boxing / unboxing primitives and using method handles / lambdas to make it work with a generically typed method. While it may be slightly better from a "clean code" perspective, I believe that in the serialisation code, such considerations should be of lesser priority than reducing/avoiding artificial overheads.
One thing that might be reasoned though is whether there isn't a benefit in extracting the body of each if-else-if into a separate method to allow the (static / JIT) compiler to potentially better / earlier optimise the byte code for the (most commonly used) fragments. In such a case, it would be easy at least to consolidate the long + date cases.

andreaskring replied:
Ok - either way is fine, i.e. we can leave it the way it is or extract the bodies into separate methods as you suggest

andreaskring commented on repository/src/main/java/org/aldica/repo/ignite/binary/AbstractCustomBinarySerializer.java:
```diff
@@ -0,0 +1,789 @@
+ *
+ * @author Axel Faust
+ */
+public abstract class AbstractCustomBinarySerializer implements BinarySerializer
```
It is fine by me to use an abstract base class and extend this for the concrete serializers, but I have a general design consideration comment, that we could consider. Using an inheritance based design makes it hard to reuse code across different concrete serializers that extend this base class. I don't know if this will ever be relevant for these serializeres, but imagine this scenario:

`SerializerX` extend `AbstractCustomBinarySerializer`
`SerializerY` extend `AbstractCustomBinarySerializer`

Assume that `SerializerZ` needs behavior from both `SerializerX` and `SerializerY`. Since multiple inheritance is not possible this could be tricky to implement in a nice way. If a compositional design approach is used instead, i.e. a design based only on interface implementation and aggregation, this would make it possible to reuse code across the different variants of the serializers in a nice way.

AFaust replied:
True, a compositional design could make it easier for some (potential) extension cases down the line. Currently, it would make the Spring configuration more verbose / cumbersome, as having fine grained optimisation properties for individual type serialisers would require us to instantiate a separate compositional delegate for the shared logic for each of the specific serialisers.
For the hypothetical situation with the X, Y and Z serialisers, I am also considering the merging of those serialisers into one implementation a valid approach as long as the type-specific handling can be cleanly delineated in the implementation. Serialisisers do not have to be the "one type - one serialiser" kind like most of the current serialisers are. The `ContentDataBinarySerializer` is an example of a combined serialiser, though it is a trivial example because both handled types are in an inheritance hierarchy and the only difference is a single additional field in the sub-class.

I would say we can and should revisit this specific design comment once we actually encounter a problematic case. As I do not consider the serialiser implementations to be part of any public API that we must keep compatible across versions, I would be perfectly fine to completely refactor the design when it becomes necessary.
That reminds me though - we have currently not defined / documented any considerations / policy with regards to what we consider to be "public API" based upon which we want to consider compatibility for semver purposes.

andreaskring replied:
Let's leave it for now and address it later on if the need should occur

andreaskring reviewed (dismissed):
Cool - thanks for making these optimizations. I have added two comments/suggestions, but everything looks good to me and it is fine by me to include the changes in this PR

AFaust committed:
Refactor tests / fix docs

- use raw serialisation without cache / memory page management overhead
  for both correctness and benchmark tests

AFaust commented:
The work in this PR did not yield sufficient improvements (at the time) compared to the computational cost to warrant inclusion. Since then [development on aldica 1.1.x](https://github.com/aldica/aldica/pull/72) has revisited this topic. By operating directly with the lower-level binary output stream and using finely grained methods to help JIT-compiler to inline/optimise the methods' bytecode, no problematic computational overhead has been observed to counteract the savings by using compressed integral values.
This PR will be closed accordingly now.
